### PR TITLE
Java 15 + Modules

### PR DESCRIPTION
**Pull request by modmuss50**
This PR fully migrates Enigma to Java 15 and modules.

I dont see a reason why we cannot go ahead with this now, especially as gradle 6.7 has a great new [toolchain](https://docs.gradle.org/6.7/userguide/toolchains.html) feature that will download java 15 if required. 

As enigma is only used in yarn that will allow gradle to automatically download java 15 if required to run engima. See https://github.com/FabricMC/fabric-loom/issues/289 as a basic exmaple of this in action.

There is currenly an issue with this where resources are not being loaded correctally when ran from idea.

### CODE REVIEW

File: enigma/src/main/java/module-info.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
+module cuchaz.enigma {
```
Comment (liach) [inline enigma/src/main/java/module-info.java:1]: Just declare this as `open module` to export everything, I suggest

Comment (modmuss50) [reply enigma/src/main/java/module-info.java:1]: Ah, didnt know that was a thing 👍 Need to look at java modules in detail a little more at somepoint. IIRC idea did most of this.

Comment (liach) [reply enigma/src/main/java/module-info.java:1]: Hmm, open is actually opening for reflection after some lookup. We still need to export all packages explicitly (but opening stuff is still not bad imo)

**Comment by modmuss50** [comment]
I am going to close this PR as its say out of date and start again for java 11.

**Comment by 2xsaiko** [comment]
Why for Java 11? Were there problems with using 15?

**Comment by modmuss50** [comment]
Yarn is now 11, going to 15 (well 16 now) would require everyone to have that installed, im not sure if thats something we can expect?

One option would be be to use gradle toolchains to download 15 for yarn, but that would mean moving all useages of engima over to invoke it via the CLI, there are a few tasks that directally invoke it atm.

**Comment by modmuss50** [comment]
After some discussion its quite clear that we cannot enforce the bleeding edge even onto just users of enigma and yarn.

**Comment by 2xsaiko** [comment]
Java 15 is barely bleeding edge, it's been out for half a year and now 16 is already out I just noticed. Also Oracle would be stupid not to extensively test even before releasing it so calling it bleeding edge even if it was just released would still be a stretch.

> would require everyone to have that installed

Well of course, but that's just how it is. Using Java 11 would require everyone to have that installed too. So what? Are there legitimate problems with using up to date Java other than people apparently being too lazy to mosey on over to the Java homepage _once_ and getting the latest Java? If we're updating anyways, might as well update to an actually up-to-date version instead of a barely newer version...

**Comment by Geolykt** [comment]
Gradle 6 breaks with Java 16 and Gradle 7 has just been released. I would personally say that it is better to move to Gradle 7 now (especially since Gradle is "just" a build system as far as I know) so those using Java 16 already do not have to work with a build system that does not even run in the first place. For yarn it is a bit different since it targets a more broader range of people, but I don't think that upgrading gradle here would impact yarn's capabilities that much, no?

**Comment by modmuss50** [comment]
Testing isnt the issue, Im confident its would be stable.

One argument is most linux package managers only have the LTS - 11
It would require people to have the default JDK be 15 or 16, in 16's case this means that everything not using gradle 7 from yeserday will fail.

I think the way forward to make this work is to use gradle's toolchains so it can auto download the latest JDK for enigma to use. This does mean some work will need to be done on yarns side to move to only using enigma via the CLI, a quick look at this  seems it's quite possible, and a good time to do it as we are moving this stuff into filament.

**Comment by 2xsaiko** [comment]
> One argument is most linux package managers only have the LTS - 11

Hmm, is that so? I didn't know that, maybe I was spoiled by running Arch for ages so I expected most other distros, at least the ones with a "middle of the road" kind of release schedule (e.g. stuff that isn't Debian stable but also not Arch) would have newer versions. That's definitely annoying, I agree.

Reminds me, I need to update my JDK packages for Gentoo to Java 16 ;)

**Comment by liach** [comment]
should we wait for 17, the next lts version?

**Comment by YanisBft** [comment]
Time to update this with java 16!

**Comment by modmuss50** [comment]
Going to close this as it’s way behind and it will be easier to start again.

most likely going to do it in 2 stages, compile against j16 and then possibly use modules later.
